packages/aws-cdk-lib/aws-dynamodb/lib/table-grants.ts

```diff
@@ -0,0 +1,178 @@
+import { ITableRef } from './dynamodb.generated';
+import * as perms from './perms';
+import * as iam from '../../aws-iam';
+import * as kms from '../../aws-kms';
+import { Stack } from '../../core';
+
+/**
+ * Construction properties for TableGrants
+ */
+export interface TableGrantsProps {
+  /**
+   * The table to grant permissions on
+   */
+  readonly table: ITableRef & iam.IResourceWithPolicy;
+
+  /**
+   * Additional regions other than the main one that this table is replicated to
+   *
+   * @default - No regions
+   */
+  readonly additionalRegions?: string[];
+
+  /**
+   * Whether this table has indexes
+   *
+   * If so, permissions are granted on all table indexes as well.
+   *
+   * @default false
+   */
+  readonly hasIndex?: boolean;
+
+  /**
+   * The encryption key for this table
+   *
+   * Required permissions will be added to the key as well.
+   *
+   * @default - No key
+   */
+  readonly encryptionKey?: kms.IKey;
+}
+
+/**
+ * A set of permissions to grant on a Table
+ */
+export class TableGrants {
+  private readonly table: ITableRef & iam.IResourceWithPolicy;
+  private readonly arns: string[] = [];
+  private readonly encryptionKey?: kms.IKey;
+
+  constructor(props: TableGrantsProps) {
```

Comment:
Same here?

Reply:
Because of the other properties (`additionalRegions` and `hasIndex`), the closest we can get to the auto-generated pattern is this:

```ts
// Actual name TBD
interface IAugmentedTable extends ITableRef {
  readonly regions?: string[];  
  readonly hasIndex?: boolean;
}

// Then in TableGrants:
_fromAugmentedTable(table: IAugmentedTable): TableGrants;
```

and make `Table`extend this new `IAugmentedTable` and, consequently, make `regions`and `hasIndex` public. Is this what you have in mind?

Another option is for the static factory to  just wrap the constructor:

```ts
_fromTableGrantsProps(props: TableGrantsProps): TableGrants;
```

Same thing for table streams.